### DISCUSSION TITLE
Is a browser extension possible?

### DISCUSSION THREAD

**Discussion by v-marinkov** [almarklein/timetagger#446] [General]
Amazing app - goed gedaan! :)
Does the API allow for a potential browser extension which would show the current running timer and allow you to start/stop timers?

**Comment by almarklein** [comment]
Interesting question! The API provides a way to create and update records at the server. The server is pretty "dump" and is not aware of what record is running. A running record, by definition, simply has its `t1` and `t2` set to the same value. The server stores such a record but does not check whether it's running or provide a way to query running records.

To provide some context, the web app has a copy of *all* records, stored in a special data structure so they can be queried very fast. Because it has all records, it also knows which are running. That's why when you open TimeTagger on both your phone and desktop, they show the same record running. It periodically asks for updates using the `GET ./updates?since=<timestamp>` API endpoint, which is pretty efficient.

The CLI takes a different approach. It stores nothing. When you stop the timer, it stops all running records. This works by querying the server for all recent record, checking which have `t1==t2`, and then updating these.

I suppose a browser extension would want to periodically check for running timers. What it could to is first get a list of running records similar to how the CLI does it, and then check for updates every 10 seconds or so.

As for starting timers. That's easy because the API allows pushing a new record. It could even load presets from the settings.
